File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    const types = new Map();
    const todayDate = new Date(today);
    const answer = [];
    
    for(const term of terms) {
        const [t, n] = term.split(' ');
        types.set(t, Number(n));
    }
    
    for(const [idx, privacy] of privacies.entries()) {
        const [date, type] = privacy.split(' ');
        const expireDate = new Date(date);
        expireDate.setMonth(expireDate.getMonth() + types.get(type));
        
        if (expireDate <= todayDate) answer.push(idx + 1);
    }
    
    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
from datetime import date
from dateutil.relativedelta import relativedelta

def solution(today, terms, privacies):
    term_dict = {}
    for term in terms:
        name, n = term.split()
        term_dict[name] = int(n)
    
    today = date(int(today.split('.')[0]), 
                 int(today.split('.')[1]), 
                 int(today.split('.')[2]))
    
    answer = []
    for i, privacy in enumerate(privacies):
        cur_date, name = privacy.split()
        year, month, day = map(int, cur_date.split('.'))
        
        # 만료 날짜는 해당 월의 마지막 날까지
        expired_date = date(year, month, day) + relativedelta(months=term_dict[name]) - relativedelta(days=1)
        
        if expired_date < today:
            answer.append(i + 1)
    
    return answer